From 5359b511f8c6e463649061151590a24a7a1f7f3f Mon Sep 17 00:00:00 2001
From: Sudzzy <originmc@outlook.com>
Date: Sat, 30 Sep 2017 17:28:35 -0400
Subject: [PATCH] Option to skip teleport safety when teleporting

People are able to abuse the way Bukkit handles teleportation across
worlds because it provides a built-in teleportation safety check.

To abuse the safety checks, players are required to get into a location
deemed to be unsafe by Bukkit. If another player is teleported to the
first, Bukkit will search upwards for a safe location, including any
bases or areas they would not normally have access to.

This only happens when teleporting between worlds.

diff --git a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
index d904700..278f2d6 100644
--- a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
+++ b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
@@ -233,4 +233,9 @@ public class PaperWorldConfig {
     private void portalSearchRadius() {
         portalSearchRadius = getInt("portal-search-radius", 128);
     }
+
+    public boolean disableTeleportationSuffocationCheck;
+    private void disableTeleportationSuffocationCheck() {
+        disableTeleportationSuffocationCheck = getBoolean("disable-teleportation-suffocation-check", false);
+    }
 }
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index b640c3a..b1182f4 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -689,7 +689,8 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
                     if (fromWorld == toWorld) {
                         entity.connection.teleport(to);
                     } else {
-                        this.server.getHandle().moveToWorld(entity, toWorld.dimension, true, to, true);
+                        // Paper - Configurable suffocation check
+                        this.server.getHandle().moveToWorld(entity, toWorld.dimension, true, to, !toWorld.paperConfig.disableTeleportationSuffocationCheck);
                     }
 
                     return true;
-- 
2.18.0

